### PR TITLE
Add open3d recipe

### PR DESCRIPTION
**Pull request by goanpeca**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by goanpeca** [comment]
Still working on this!

**Comment by Tobias-Fischer** [comment]
@conda-forge-admin please restart ci

**Comment by Tobias-Fischer** [comment]
Gosh there are lots of vendored packages in there that we need to fix up. @goanpeca  any chance could give me access to your fork?

**Comment by goanpeca** [comment]
Still working on this!

**Comment by Tobias-Fischer** [comment]
@goanpeca - I think this should be nearly ready for review, we just need to wait for upstream to release 0.18.0 which hopefully should be soon

**Comment by goanpeca** [comment]
Awesome @Tobias-Fischer ! Thanks 🙌🏼

**Comment by Tobias-Fischer** [comment]
@conda-forge/help-c-cpp 

This PR is ready for review. As noted in the meta.yaml, there are quite a few packages that are vendored by open3d - I attempted to use as many system libraries as possible, but some of them were either not available on conda-forge, only incompatible versions (tbb), or led to build errors that I could not fix etc.; however, I made sure to add all third-party licenses to the meta.yaml and these libraries are not contained in the package (static builds and upstream has taken care).

**Comment by Tobias-Fischer** [comment]
@goanpeca - could you please remove the "Draft" label?

**Comment by Tobias-Fischer** [comment]
Hi @conda-forge/help-c-cpp @goanpeca - a kind reminder :)

**Comment by carterbox** [comment]
Restarting build so I can download the artifacts.

File: recipes/Open3D/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    - liblapack
+    - liblapacke
+    # - nanoflann  # TODO: Does not export cmake config in conda-forge; https://github.com/conda-forge/nanoflann-feedstock/issues/4
+    # - tinygltf   # TODO: Does not export cmake config in conda-forge
+    # - libcurl    # does not like the conda-forge version, results in linker error
+    # - openssl    # does not like the conda-forge version, results in linker error
+    # - tbb-devel  # conda-forge version too new
+    # - imgui      # Not on conda-forge
+    # - libzlf     # Not on conda-forge
+  run:
```
Comment (carterbox) [inline recipes/Open3D/meta.yaml:82]: Please address the overdepending warnings for the non-python packages in the build logs:

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=853669&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=4227

File: recipes/Open3D/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    - libcblas
+    - liblapack
+    - liblapacke
+    # - nanoflann  # TODO: Does not export cmake config in conda-forge; https://github.com/conda-forge/nanoflann-feedstock/issues/4
+    # - tinygltf   # TODO: Does not export cmake config in conda-forge
+    # - libcurl    # does not like the conda-forge version, results in linker error
+    # - openssl    # does not like the conda-forge version, results in linker error
+    # - tbb-devel  # conda-forge version too new
+    # - imgui      # Not on conda-forge
+    # - libzlf     # Not on conda-forge
```
Comment (carterbox) [inline recipes/Open3D/meta.yaml:81]: Please add additional recipes to this PR for imgui and zlf. They both have multiple stable releases to choose from and already have USE_SYSTEM_* options implemented in the CMakeLists for open3d.

**Review by carterbox — CHANGES_REQUESTED** [review]

**Comment by carterbox** [comment]
This package is shipping libOpen3D.so with cmake, pkgconfig, and headers. It needs run_exports and should preferably be split into separate shared library and python parts OR the development stuff should be removed until a later time when someone requests the ability to link to libOpen3D directly.

**Comment by Tobias-Fischer** [comment]
Thanks for the detailed review @carterbox, I appreciate it!

That said, I won't have the capacity to incorporate your suggestions anytime soon; there are quite a few things to change. @goanpeca are you happy to give this a shot?

PS: For anyone looking,

File: recipes/Open3D/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    - {{ cdt('libxau') }}                    # [linux]
+    - {{ cdt('libxi-devel') }}               # [linux]
+    - python                                 # [build_platform != target_platform]
+    - pip                                    # [build_platform != target_platform]
+    - cross-python_{{ target_platform }}     # [build_platform != target_platform]
+  host:
+    - setuptools
+    - python
+    - pip
+    - libboost-devel
```
[the remark this answers is not in the document]
Comment (Tobias-Fischer) [reply recipes/Open3D/meta.yaml]: boost not needed anymore; removed.

[the remark this answers is not in the document]
Comment (Tobias-Fischer) [reply recipes/Open3D/meta.yaml]: boost was required after all for a third-party dependency, but is statically linked in, so have added to `ignore_run_exports`.

File: recipes/Open3D/meta.yaml (hunk 2 above)
Comment (Tobias-Fischer) [reply recipes/Open3D/meta.yaml]: Have added.

**Comment by Tobias-Fischer** [comment]
> This package is shipping libOpen3D.so with cmake, pkgconfig, and headers. It needs run_exports and should preferably be split into separate shared library and python parts OR the development stuff should be removed until a later time when someone requests the ability to link to libOpen3D directly.

Have added `run_exports`. Is the split into multiple packages really necessary? If so, is there an example package that shows how to best do that?

File: recipes/Open3D/meta.yaml (hunk 1 above)
Comment (Tobias-Fischer) [reply recipes/Open3D/meta.yaml:89]: Done

**Comment by Tobias-Fischer** [comment]
@carterbox @conda-forge/help-python-c  this is ready for another review :).

**Comment by Tobias-Fischer** [comment]
Hi @carterbox @conda-forge/help-python-c - a quick reminder to please take a look :)

File: recipes/liblzf/build.sh [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+rm -rf build || true
+mkdir build
+cd build
+
+cmake ${SRC_DIR} ${CMAKE_ARGS} \
+    -DBUILD_SHARED_LIBS=ON
+
+cmake --build . --config Release -- -j$CPU_COUNT
+cmake --build . --config Release --target install
```
Comment (carterbox) [inline recipes/liblzf/build.sh:11]: Why is the source being included in the package?

File: recipes/Open3D/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    - 3rdparty/uvatlas/LICENSE_directxheaders
+    - 3rdparty/uvatlas/LICENSE_directxmath
+    - 3rdparty/uvatlas/LICENSE_uvatlas
+    - 3rdparty/parallelstl/LICENSE
+    - 3rdparty/possionrecon/LICENSE
+    - 3rdparty/rply/LICENSE
+    - 3rdparty/curl/LICENSE
+    - 3rdparty/boringssl/LICENSE
+    - 3rdparty/embree/LICENSE
+    - 3rdparty/liblzf/LICENSE
```
Comment (carterbox) [inline recipes/Open3D/meta.yaml:117]:
```suggestion
```
You're now dynamically linking, so this is unneeded?

File: recipes/Open3D/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    - fix-unzip.patch
+    - fix-curl.patch
+    - fix-error.patch
+    - fix-find-imgui.patch
+
+build:
+  number: 0
+  skip: True  # [win]
+  ignore_run_exports_from:
+    - libboost-devel
```
Comment (carterbox) [inline recipes/Open3D/meta.yaml:22]:
```
Could NOT find Boost: Found unsuitable version "1.82.0", but required is exact version "1.84.0" (found $PREFIX/include, )
```
Your package requires a specific version of boost, and it's not getting that version. Maybe that's why you have to ignore run exports?

File: recipes/Open3D/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    - libcblas
+    - minizip
+    - msgpack-c
+    - msgpack-cxx
+  run_exports:
+    - {{ pin_subpackage('open3d', max_pin='x.x.x') }}
```
Comment (carterbox) [inline recipes/Open3D/meta.yaml:28]: Do you want help splitting this package into dev and non-dev packages or you are OK will all users getting all artifacts and dependencies?

**Review by carterbox — CHANGES_REQUESTED** [review]

File: recipes/Open3D/meta.yaml (hunk 7 above)
Comment (Tobias-Fischer) [reply recipes/Open3D/meta.yaml:28]: I am ok with all users getting all artefacts/dependencies.

File: recipes/liblzf/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+file(READ lzf.h lzf_content)
+if(lzf_content MATCHES ".*#define LZF_VERSION (0x[0-9A-Fa-f]+).*")
+	# CMake 3.13+ is needed to evaluate 0x hexadecimal input values
+	math(EXPR API_VERSION_MAJOR "${CMAKE_MATCH_1} >> 8")
+	math(EXPR API_VERSION_MINOR "${CMAKE_MATCH_1} & 255")
+	set(API_VERSION "${API_VERSION_MAJOR}.${API_VERSION_MINOR}")
+	message(STATUS "liblzf API ${API_VERSION}")
+else()
+	message(FATAL_ERROR "cannot extract API version from lzf.h")
+endif()
```
[the remark this answers is not in the document]
Comment (Tobias-Fischer) [reply recipes/liblzf/CMakeLists.txt:43]: I have simply copied the Debian CMakeLists.txt that they use, assuming that they know what they are doing :).

https://sources.debian.org/src/liblzf/3.6-4/

File: recipes/liblzf/build.sh (hunk 4 above)
Comment (Tobias-Fischer) [reply recipes/liblzf/build.sh:11]: Good catch, fixed.

File: recipes/Open3D/meta.yaml (hunk 6 above)
Comment (Tobias-Fischer) [reply recipes/Open3D/meta.yaml]: This is an issue in vtk (see next line "Could not find the VTK package due to a missing dependency: Boost"). The vtk recipe uses `libboost-headers` only (version 1.84.0, as there is no global pin on libboost-headers), but then their cmake scripts require the exact version of boost when projects depend on vtk.

I will try again just depending on `libboost-headers` and see how we go, but I think I've encountered issues with that previously.

File: recipes/liblzf/CMakeLists.txt (hunk 8 above)
[the remark this answers is not in the document]
Comment (Tobias-Fischer) [reply recipes/liblzf/CMakeLists.txt:43]: Also note that I have no clue about this library - I simply added it as it is a dependency of Open3D and you had suggested it :)

File: recipes/Open3D/meta.yaml (hunk 6 above)
Comment (Tobias-Fischer) [reply recipes/Open3D/meta.yaml]: `libboost-headers` works just fine :)

**Comment by Tobias-Fischer** [comment]
Hi @carterbox @conda-forge/help-python-c - this should be all good now, I've addressed all your feedback. Many thanks for taking the time to review!

**Comment by Tobias-Fischer** [comment]
Thanks a lot for the thorough review @carterbox :)

**Pull request merged**